### ISSUE TITLE
Invalid identifier as dict key in 802.1x support

### ISSUE THREAD

**Issue by pcahyna** [linux-system-roles/network#189]
As mentioned in the discussion in #159, the 802.1x support adds a `802.1x` key to a dict in the user interface. While this is  is in principle possible, I believe this is confusing and it could be better to stick to a standard of having only valid identifiers as dict keys. The reason is that in Jinja2 one often uses a dot to dereference dict keys, so dict keys containing a dot are confusing - it looks like another layer of dict is being introduced. Also, when the dict key is not a valid identifier, one can not use the usual dot notation to reference it, but one has to use the bracket notation ( `dict['802.1x']` ), leading to a potential surprise and a less uniform syntax.
Of course, there is a precedent of `features` which use plenty of keys that are invalid identifiers - contain dashes. It is a subject for a standardization discussion whether in the future we want to avoid such names or we don't care. The current standards are not very clear on this topic:
https://github.com/oasis-roles/meta_standards#naming-things says just "All YAML or Python files, variables, arguments, repositories, and other such names should follow standard Python naming conventions of being in snake_case_naming_schemes." which does not specify whether dictionary keys are among "other such names". We then have "Do not use special characters other than underscore in variable names, even if YAML/JSON allow them." which speaks only about variable names (where it is more important than for dictionary keys).

**Comment by pcahyna** [comment]
I tried the following dictionary keys:
```yaml
      802.1x
      802-1x
      802_1x
      ieee802.1x
      ieee802-1x
      ieee802_1x
```
Only the last one was usable with the dot notation to access it (`dict.ieee802_1x`).

**Comment by tyll** [comment]
It is not very convincing that that it is necessary to make dictionary keys valid python identifiers. Especially if they are derived from other tools (like the features from ethtool), it seems to be more valid to actually use the normal value instead of inventing different names.

Regarding 802.1x, I see some value in moving this in a more generic authentication dictionary like

```yaml
authentication:
    mode: 802.1x|wpa-psk
    ....
```

see #157. What is your opinion about this @pcahyna @jaedolph? However, some more keys for the authentication/802.1x setting are also using dashes instead of underscores such as `private-key`. IMHO it is valid for users to use the bracket notation for this to benefit from being able to use the same settings name in nmcli and the system role.

Also, if we need to watch for these naming conventions, it would be beneficial to have a lint checker for this in the CI to avoid surprises.

**Comment by tyll** [comment]
> The OASIS project also ran into [this issue](https://github.com/ansible/galaxy/issues/775) with Galaxy, which is similar, and also caused breaking changes for our users.
> 
> Given these pitfalls, I recommend to simply use python identifier names for _everything_, from variables to filenames and so on. This does have the downside of causing some consistency issues, where you can't name things exactly like you'd want to, but the possibility of breaking changes for users is more important in my humble opinion.

I just realised this: The roles are prefixed with `linux-system-roles` which is also an invalid identifier. Do you think this will cause problems as well? The Galaxy issue was about problems installing roles with dashes in their name.

**Comment by pcahyna** [comment]
I indeed think that this may be a problem in the future. I think that one would not be able to create such a prefix in Galaxy anymore. @t-woerner could you please share your observations on this topic?

**Comment by jaedolph** [comment]
@tyll @pcahyna 

> I tried the following dictionary keys:
> 
> ```yaml
>       802.1x
>       802-1x
>       802_1x
>       ieee802.1x
>       ieee802-1x
>       ieee802_1x
> ```
> 
> Only the last one was usable with the dot notation to access it (`dict.ieee802_1x`).

I think ieee802_1x is a good identifier for the 802.1x parameter.

> 
> Regarding 802.1x, I see some value in moving this in a more generic authentication dictionary like
> 
> ```yaml
> authentication:
>     mode: 802.1x|wpa-psk
>     ....
> ```

In my opinion this might be a bit confusing for users as I don't think wpa-psk can be used for wired connections? I feel like it would make more sense to keep this under the "wireless" parameter and change the 802.1x key to ieee802_1x. This also corresponds to how the options are laid out in NetworkManager. Happy to use `authentication` though if we think this would be preferable

**Comment by jaedolph** [comment]
This PR should fix this  #157

**Comment by tyll** [comment]
@pcahyna do we need to change the dictionary keys for ethtool features, too? When #112 and #113 will be implemented, similar options might get introduced.

**Comment by pcahyna** [comment]
@tyll the ethtool features keys have been already released, right? We then can not change them, but we should not introduce new keys which do not conform to the new rules.

(I see that you asked about dashes in names at that time, https://github.com/linux-system-roles/network/issues/17#issuecomment-499660548 , I should have had more foresight, sigh! )

**Comment by tyll** [comment]
> @tyll the ethtool features keys have been already released, right? We then can not change them, but we should not introduce new keys which do not conform to the new rules.

IMHO we can still deprecate the keys by change the code to accept also keys with underscore instead of dashes, warn if dashes are used and update the docs to mention only the version with underscore. Then it will at least be uniform for new users.

**Comment by pcahyna** [comment]
@tyll that looks like the right approach to me as well.

**Comment by tyll** [comment]
I created more focused issues on the remaining tasks. Please add more if something is missing. The issue reported here was fixed in #202, therefore I close this issue.
